### PR TITLE
Implementation of The Visualisation Binary Protocol

### PR DESCRIPTION
**Pull request by lolczak**
### Pull Request Description
An implementation of push-side of the visualization binary protocol.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/style-guides/scala.md), [Java](https://github.com/luna/enso/blob/master/doc/style-guides/java.md), [Rust](https://github.com/luna/enso/blob/master/doc/style-guides/rust.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/style-guides/haskell.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/boot/LanguageServerComponent.scala [hunk 1]
```diff
@@ -83,3 +91,20 @@ class LanguageServerComponent(config: LanguageServerConfig)
...
+object LanguageServerComponent {
+
+  /**
+    * A running server context.
+    *
+    * @param mainModule a main module containing all components of the server
+    * @param rpcBinding a http binding for rpc protocol
+    * @param dataBinding a http binding for data protocol
+    */
+  case class ServerCtx(
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/boot/LanguageServerComponent.scala:104]: This should be called `ServerContext`. The type shouldn't be abbreviated, but usages can be.

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  BinaryDecoder,
+  BinaryEncoder,
+  DecodingFailure
+}
+
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * A web socket server realizing binary protocol.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala:31]: This doesn't make sense in english.

```suggestion
  * A web socket server using a binary protocol.
```

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  *
+  * @param decoder an inbound packets decoder
+  * @param encoder an outbound packets encoder
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala:34]:
```suggestion
  * @param decoder a decoder for inbound packets
  * @param encoder an encoder for outbound packets
```

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  * @param factory creates connection front controller per a single connection
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala:35]: I don't understand what this one is trying to say. In general the doc comments should describe what they're _for_, not what they _are_.

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  implicit val materializer: Materializer
+) extends LazyLogging {
+
+  implicit val ec: ExecutionContext = system.dispatcher
+
+  private val route: Route =
+    extractClientIP {
+      case RemoteAddress.Unknown =>
+        complete(
+          StatusCodes.InternalServerError -> "Set akka.http.server.remote-address-header to on"
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala:58]: Overly long line.

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+  private def createInboundFlow(
+    frontController: ActorRef,
+    ip: RemoteAddress.IP
+  ): Sink[Message, NotUsed] = {
+    Flow[Message]
+      .mapConcat[BinaryMessage] {
+        case msg: TextMessage =>
+          logger.warn(
+            s"Received text message $msg inside binary protocol [$ip]"
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala:122]:
```suggestion
            s"Received text message $msg over the data connection [$ip]"
```

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala [hunk 7]
```diff
@@ -0,0 +1,46 @@
+package org.enso.languageserver.http.server
+
+import akka.actor.ActorRef
+
+/**
+  * A web socket control protocol used to manage connection created by
+  * [[BinaryWebSocketServer]].
+  */
+object WebSocketControlProtocol {
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala:9]: Nothing about this name mentions that it's to do only with the `BinaryWebSocketServer`. It should be namespaced to indicate this, or renamed.

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    */
+  sealed trait WsEvent
+
+  /**
+    * Base trait for web socket commands.
+    */
+  sealed trait WsCommand
+
+  /**
+    * Signals that connection has been closed.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala:22]:
```suggestion
    * Signals that a connection has been closed.
```

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+    * Signals connection failure.
+    *
+    * @param throwable a throwable
+    */
+  case class ConnectionFailed(throwable: Throwable) extends WsEvent
+
+  /**
+    * Signals that connection has been established.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala:34]:
```suggestion
    * Signals that a connection has been established.
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/BinaryConnectionController.scala [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    with UnhandledLogging {
+
+  override def receive: Receive =
+    connectionEndHandler orElse connectionNotEstablished
+
+  private def connectionNotEstablished: Receive = {
+    case OutboundStreamEstablished(outboundChannel) =>
+      log.info(s"Connection established [$clientIp]")
+      unstashAll()
+      context.become(
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/BinaryConnectionController.scala:57]: This line is overly long.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/BinaryConnectionControllerFactory.scala [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package org.enso.languageserver.protocol.data
+
+import akka.actor.{ActorRef, ActorSystem, Props}
+import akka.http.scaladsl.model.RemoteAddress
+import org.enso.languageserver.http.server.ConnectionControllerFactory
+
+/**
+  * A factory of binary connection controllers.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/BinaryConnectionControllerFactory.scala:8]:
```suggestion
  * A factory for binary connection controllers.
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/InboundMessageDecoder.scala [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+}
+import org.enso.languageserver.util.binary.DecodingFailure.{
+  DataCorrupted,
+  EmptyPayload,
+  GenericDecodingFailure
+}
+import org.enso.languageserver.util.binary.{BinaryDecoder, DecodingFailure}
+
+/**
+  * A decoder of [[InboundMessage]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/InboundMessageDecoder.scala:17]:
```suggestion
  * A decoder for an [[InboundMessage]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.google.flatbuffers.FlatBufferBuilder
+import org.enso.languageserver.protocol.data.util.EnsoUUID
+
+object EnsoUuidFactory {
+
+  /**
+    * Creates EnsoUUID inside a [[FlatBufferBuilder]].
+    *
+    * @param uuid a uuid do serialize
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala:13]:
```suggestion
    * @param uuid a uuid to serialize
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  def create(uuid: UUID)(implicit builder: FlatBufferBuilder): Int = {
+    EnsoUUID.createEnsoUUID(
+      builder,
+      uuid.getLeastSignificantBits,
+      uuid.getMostSignificantBits
+    )
+  }
+
+  /**
+    * Creates EnsoUUID inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala:26]:
```suggestion
    * Creates an [[EnsoUUID]] inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    *
+    * @param uuid a uuid do serialize
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala:28]:
```suggestion
    * @param uuid a uuid to serialize
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala [hunk 16]
```diff
@@ -0,0 +1,53 @@
+package org.enso.languageserver.protocol.data.factory
+
+import com.google.flatbuffers.FlatBufferBuilder
+import org.enso.languageserver.protocol.data.util.Error
+
+object ErrorFactory {
+
+  /**
+    * Creates ReceivedCorruptedDataError inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala:9]:
```suggestion
    * Creates a ReceivedCorruptedDataError inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    implicit builder: FlatBufferBuilder
+  ): Int =
+    Error.createError(
+      builder,
+      1,
+      builder.createString("Received corrupted data")
+    )
+
+  /**
+    * Creates ReceivedEmptyPayloadError inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala:24]:
```suggestion
    * Creates a ReceivedEmptyPayloadError inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+    implicit builder: FlatBufferBuilder
+  ): Int =
+    Error.createError(
+      builder,
+      2,
+      builder.createString("Received empty payload in the inbound message")
+    )
+
+  /**
+    * Creates ServiceError inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala:39]:
```suggestion
    * Creates a ServiceError inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/OutboundMessageFactory.scala [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.UUID
+
+import com.google.flatbuffers.FlatBufferBuilder
+import org.enso.languageserver.protocol.data.envelope.OutboundMessage
+import org.enso.languageserver.protocol.data.util.EnsoUUID
+
+object OutboundMessageFactory {
+
+  /**
+    * Creates [[OutboundMessage]] inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/OutboundMessageFactory.scala:12]:
```suggestion
    * Creates an [[OutboundMessage]] inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/SessionInitResponseFactory.scala [hunk 20]
```diff
@@ -0,0 +1,19 @@
+package org.enso.languageserver.protocol.data.factory
+
+import com.google.flatbuffers.FlatBufferBuilder
+import org.enso.languageserver.protocol.data.session.SessionInitResponse
+
+object SessionInitResponseFactory {
+
+  /**
+    * Creates [[SessionInitResponse]] inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/SessionInitResponseFactory.scala:9]:
```suggestion
    * Creates a [[SessionInitResponse]] inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+import org.enso.languageserver.protocol.data.executioncontext
+import org.enso.languageserver.runtime.VisualisationProtocol.{
+  VisualisationContext,
+  VisualisationUpdate
+}
+
+object VisualisationUpdateFactory {
+
+  /**
+    * Creates [[VisualisationUpdate]] inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala:13]:
```suggestion
    * Creates a [[VisualisationUpdate]] inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+        .createDataVector(builder, update.data)
+    executioncontext.VisualisationUpdate.createVisualisationUpdate(
+      builder,
+      ctx,
+      data
+    )
+  }
+
+  /**
+    * Creates [[VisualisationContext]] inside a [[FlatBufferBuilder]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala:34]:
```suggestion
    * Creates a [[VisualisationContext]] inside a [[FlatBufferBuilder]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/runtime/VisualisationProtocol.scala [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+    * @param expressionId an expression identifier
+    */
+  case class VisualisationContext(
+    visualisationId: UUID,
+    contextId: UUID,
+    expressionId: UUID
+  )
+
+  /**
+    * Event signaling visualisation update.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/runtime/VisualisationProtocol.scala:21]:
```suggestion
    * An event signaling a visualisation update.
```

File: engine/language-server/src/main/scala/org/enso/languageserver/util/binary/DecodingFailure.scala [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+  * Base trait signaling decoding failure.
+  */
+sealed trait DecodingFailure
+
+object DecodingFailure {
+
+  /**
+    * Signals empty payload in byte stream.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/util/binary/DecodingFailure.scala:11]:
```suggestion
    * Signals an empty payload in the byte stream.
```

File: engine/language-server/src/main/scala/org/enso/languageserver/util/binary/DecodingFailure.scala [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+    */
+  case object EmptyPayload extends DecodingFailure
+
+  /**
+    * Signals that data is corrupted.
+    */
+  case object DataCorrupted extends DecodingFailure
+
+  /**
+    * Represents undefined decoding failure.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/util/binary/DecodingFailure.scala:21]:
```suggestion
    * Represents an undefined decoding failure.
```

File: engine/language-server/src/main/schema/envelope.fbs [hunk 26]
```diff
@@ -0,0 +1,27 @@
+include "util.fbs";
```
Comment (iamrecursion) [inline engine/language-server/src/main/schema/envelope.fbs:1]: I'd like to see descriptive comments added to all of the IDL files.

File: engine/language-server/src/main/schema/envelope.fbs [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+  ERROR: org.enso.languageserver.protocol.data.util.Error,
+  SESSION_INIT_RESPONSE: org.enso.languageserver.protocol.data.session.SessionInitResponse,
+  VISUALISATION_UPDATE: org.enso.languageserver.protocol.data.executioncontext.VisualisationUpdate
+}
+
+table OutboundMessage {
+  requestId: org.enso.languageserver.protocol.data.util.EnsoUUID (required);
+  correlationId: org.enso.languageserver.protocol.data.util.EnsoUUID;
+  payload: OutboundPayload (required);
+}
```
Comment (iamrecursion) [inline engine/language-server/src/main/schema/envelope.fbs:27]: Please also add a blank line at the end of all the IDL files.

File: lib/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerBootLoader.scala [hunk 28]
```diff
@@ -44,34 +44,32 @@ class LanguageServerBootLoader(
...
-        descriptor.networkConfig.maxPort
-      )
+      val rpcPort  = findPort()
+      var dataPort = findPort()
+      while (dataPort == rpcPort) {
+        dataPort = findPort()
+      }
       log.info(
-        s"Found a socket for the language server [${descriptor.networkConfig.interface}:$port]"
+        s"Found sockets for the language server [rpc:${descriptor.networkConfig.interface}:$rpcPort, data:${descriptor.networkConfig.interface}:$dataPort]"
```
Comment (iamrecursion) [inline lib/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerBootLoader.scala:53]: Overly long line. Please wrap.

File: engine/language-server/src/main/scala/org/enso/languageserver/boot/LanguageServerComponent.scala (hunk 1 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/boot/LanguageServerComponent.scala]: Done

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala (hunk 4 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala:35]: corrected

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala (hunk 7 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/http/server/WebSocketControlProtocol.scala:9]: fixed

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/BinaryConnectionController.scala (hunk 10 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/BinaryConnectionController.scala:57]: fixed

File: engine/language-server/src/main/schema/envelope.fbs (hunk 27 above)
Comment (lolczak) [reply engine/language-server/src/main/schema/envelope.fbs]: done

File: engine/language-server/src/main/schema/envelope.fbs (hunk 26 above)
Comment (lolczak) [reply engine/language-server/src/main/schema/envelope.fbs:1]: done

File: lib/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerBootLoader.scala (hunk 28 above)
Comment (lolczak) [reply lib/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerBootLoader.scala]: done

File: engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala [hunk 29]
```diff
@@ -6,33 +6,35 @@ import java.net.URI
...
 import org.graalvm.polyglot.Context
 import org.graalvm.polyglot.io.MessageEndpoint
 
 import scala.concurrent.duration._
 
 /**
-  * A main module containing all components of th server.
+  * A main module containing all components of the server.
   *
   * @param serverConfig a server config
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala:39]:
```suggestion
  * @param serverConfig configuration for the language server
```

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketControlProtocol.scala [hunk 30]
```diff
@@ -0,0 +1,46 @@
+package org.enso.languageserver.http.server
+
+import akka.actor.ActorRef
+
+/**
+  * A web socket control protocol used to manage connections created by
+  * [[BinaryWebSocketServer]].
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketControlProtocol.scala:7]:
```suggestion
  * a [[BinaryWebSocketServer]].
```

File: engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala [hunk 31]
```diff
@@ -0,0 +1,178 @@
...
+        }
+    }
+
+  /**
+    * Binds this server instance to a given port and interface, allowing
+    * future connections.
+    *
+    * @param interface the interface to bind to.
+    * @param port the port to bind to.
+    * @return a server binding object.
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/http/server/BinaryWebSocketServer.scala:77]:
```suggestion
    * @return a representation of the binding state of the server.
```

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+import org.enso.languageserver.protocol.data.util.EnsoUUID
+
+object EnsoUuidFactory {
+
+  /**
+    * Creates EnsoUUID inside a [[FlatBufferBuilder]].
+    *
+    * @param uuid a uuid to serialize
+    * @param builder a builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala:15]: These two need to be more detailed.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+      uuid.getMostSignificantBits
+    )
+  }
+
+  /**
+    * Creates an [[EnsoUUID]] inside a [[FlatBufferBuilder]].
+    *
+    * @param uuid a uuid to serialize
+    * @param builder a builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/EnsoUuidFactory.scala:30]: As do these two.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala [hunk 34]
```diff
@@ -0,0 +1,53 @@
...
+import com.google.flatbuffers.FlatBufferBuilder
+import org.enso.languageserver.protocol.data.util.Error
+
+object ErrorFactory {
+
+  /**
+    * Creates a ReceivedCorruptedDataError inside a [[FlatBufferBuilder]].
+    *
+    * @param builder a flat buffer builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala:12]: These two should be more detailed.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala [hunk 35]
```diff
@@ -0,0 +1,53 @@
...
+      builder,
+      1,
+      builder.createString("Received corrupted data")
+    )
+
+  /**
+    * Creates a ReceivedEmptyPayloadError inside a [[FlatBufferBuilder]].
+    *
+    * @param builder a flat buffer builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala:27]: More detail.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+      builder,
+      2,
+      builder.createString("Received empty payload in the inbound message")
+    )
+
+  /**
+    * Creates a ServiceError inside a [[FlatBufferBuilder]].
+    *
+    * @param builder a flat buffer builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/ErrorFactory.scala:42]: Here too.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/OutboundMessageFactory.scala [hunk 37]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+    * Creates an [[OutboundMessage]] inside a [[FlatBufferBuilder]].
+    *
+    * @param requestId a request id
+    * @param maybeCorrelationId a optional correlation id
+    * @param payloadType a payload type
+    * @param payload a message payload
+    * @param builder a flat buffer builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/OutboundMessageFactory.scala:19]: These should be more detailed.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/SessionInitResponseFactory.scala [hunk 38]
```diff
@@ -0,0 +1,19 @@
...
+import com.google.flatbuffers.FlatBufferBuilder
+import org.enso.languageserver.protocol.data.session.SessionInitResponse
+
+object SessionInitResponseFactory {
+
+  /**
+    * Creates a [[SessionInitResponse]] inside a [[FlatBufferBuilder]].
+    *
+    * @param builder a flat buffers builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/SessionInitResponseFactory.scala:12]: More detail.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala [hunk 39]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+object VisualisationUpdateFactory {
+
+  /**
+    * Creates a [[VisualisationUpdate]] inside a [[FlatBufferBuilder]].
+    *
+    * @param update a visualisation update
+    * @param builder a flat buffers builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala:17]: More detail.

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala [hunk 40]
```diff
@@ -0,0 +1,53 @@
...
+      data
+    )
+  }
+
+  /**
+    * Creates a [[VisualisationContext]] inside a [[FlatBufferBuilder]].
+    *
+    * @param ctx a VisualisationContext
+    * @param builder a flat buffers builder
+    * @return offset
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/data/factory/VisualisationUpdateFactory.scala:38]: More detail.

File: engine/language-server/src/main/schema/envelope.fbs (hunk 26 above)
Comment (iamrecursion) [reply engine/language-server/src/main/schema/envelope.fbs:1]: I want to see more detail. The components of each message should be commented.

File: lib/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerService.scala [hunk 41]
```diff
@@ -22,12 +22,12 @@ trait LanguageServerService[F[+_, +_]] {
     *
     * @param clientId a requester id
     * @param project a project to start
-    * @return either failure or socket
+    * @return either failure or sockets
```
Comment (iamrecursion) [inline lib/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerService.scala:25]: More detail.

**Pull request merged**